### ISSUE TITLE
Custom endpoint not available anymore

### ISSUE THREAD

**Issue by alpipego** [ibericode/koko-analytics#70]
Here's my setup: 
I install WordPress via composer into its own directory `wp` and `wp-content` is on the same level as `wp`.

```
public
  |-- wp
  |-- wp-content
  |-- index.php
  |-- wp-config.php
  |-- koko-analytics-collect.php
```

The way the custom endpoint is setup since version 1.0.12 is that it looks for the `koko-analytics-collect.php` within the `wp` dir, by using `ABSPATH` and `site_url()` (instead of `home_url()`).

I understand that my setup is not 100% ordinary, but then again it's not so uncommon to install wp via composer and give it it's own directory (there's even a codex entry on it).

In short: The way it was in 1.0.11 worked for me, after I copied the `koko-analytics-collect.php` into the web-root, with 1.0.12 it looks for that file in the wrong directory.

**Comment by dannyvankooten** [comment]
Hey @alpipego,

So if I understand correctly you were already using the custom endpoint file by copying it manually and then setting the `KOKO_ANALYTICS_USE_CUSTOM_ENDPOINT` constant to `true`, correct?

So if we were to use `home_url` again instea of `site_url` (which makes sense to me, I probably shouldn't have changed that), things whould start working again for you. Correct?

**Comment by alpipego** [comment]
Yes, changing the `site_url` back to `home_url` works for me.

**Comment by ouun** [comment]
Hi @alpipego,
I have the exact same issue on multiple roots/bedrock sites. WordPress is required as Composer package and installs to `/wp` as `ABSPATH`. Updating WordPress via Composer removes the whole `/wp` folder and with that the Koko endpoint. Therefore I am missing all data until I open the Analytics Dashboard and the file is recreated.

I see that the changes suggested here have not landed, yet. But I strongly feel that `ABSPATH` is not the best position to store the file. `WP_CONTENT_DIR` could be a better alternative to save the file, at the same level as `advanced-cache.php` and other drop-ins are placed.

Kind regards

**Comment by alpipego** [comment]
@ouun this has been a while. But you can copy the [drop-in collect file](https://github.com/ibericode/koko-analytics/blob/master/koko-analytics-collect.php) to your public root directory (change the paths in the file accordingly) and set the `KOKO_ANALYTICS_USE_CUSTOM_ENDPOINT` to `true`.

**Comment by ouun** [comment]
Danke @alpipego! Just tried that and the endpoint url changed to `/wp/wp-admin/admin-ajax.php?action=koko_analytics_collect`, what works but is not optimized. 
In the settings of Koko Analytics I find the hint, that the optimized endpoint file should be in the "WordPress root directory" which is for Bedrock `/wp/koko-analytics-collect.php`, and that is exactly the issue. It should not be part of the Composer package of WordPress within `/wp`. 

The "Endpoint_Installer" class only checks for the existing file via `file_exists( ABSPATH . '/koko-analytics-collect.php' )`, which again uses `ABSPATH` - what is the path to WordPress (here as Composer package) and gets completely replaced with every update. 

What seems important to me is, that `ABSPATH` is the absolute path to the WordPress folder, where `wp-load.php` is based. IMO it is not the place for endpoint files that should go to the public folder, which is `WP_CONTENT_DIR`.

However I am wondering if you are open for a PR, as `KOKO_ANALYTICS_USE_CUSTOM_ENDPOINT` could be set with a variable which defines the path to the `koko-analytics-collect.php`file. So in my case I would simply set:

```
# true OR path to the endpoint
define('KOKO_ANALYTICS_USE_CUSTOM_ENDPOINT', WP_UPLOADS_DIR . '/koko-analytics-collect.php');
```

I am open to create a PR if that would be of benefit for you, too.

Thanks!

**Comment by alpipego** [comment]
Ah, yes. Sorry. See, I said it's been a while. I have a fork of this https://github.com/alpipego/koko-analytics/blob/fix-1.0.26/src/class-script-loader.php#L65-L72 with some added filters: https://github.com/alpipego/koko-analytics/commit/6adfa94c75db538dadf8dbc87e4600954126b421

@dannyvankooten I never opened a PR for this. Is this something you would merge?

**Comment by dannyvankooten** [comment]
Hi @alpipego and @ouun,

The issue is that a lot of websites restrict web access to files in /wp-content/, especially PHP files. That's why we went for the root directory as the preferred location for the file.

After 9a50f9a260f70b6d3b052ad3d71de10092b2cc36 you have two options:

1. Disable the custom endpoint entirely by defining a constant:
```php
define('KOKO_ANALYTICS_CUSTOM_ENDPOINT', false );
```

2. [Manually create the endpoint file](https://github.com/ibericode/koko-analytics/blob/master/koko-analytics-collect.php) somewhere accessible through your web server where it isn't deleted, fix all file references and then point Koko to your custom endpoint:

```php
define('KOKO_ANALYTICS_CUSTOM_ENDPOINT', '/path-to-endpoint-file.php' );
```

**Comment by dannyvankooten** [comment]
One other thing I am thinking of to allow the use of the default custom endpoint file and health checker is to have the endpoint installer re-install and re-test after every WordPress update. I am not sure whether Bedrock fires the various "after core update" hooks though... Do you know?

**Comment by ouun** [comment]
> The issue is that a lot of websites restrict web access to files in /wp-content/, especially PHP files. That's why we went for the root directory as the preferred location for the file.

I see. Makes sense. I am very happy with the filter. Thanks a lot!

> I am not sure whether Bedrock fires the various "after core update" hooks though... Do you know?

I am not sure either. However even for people who do not use roots/bedrock it is quite common to use other solutions for installing WordPress as Composer dependency. And many of them (including me) deactivate the health tests, so I would not count on that. :)
